src/main/java/spoon/reflect/visitor/chain/CtQueryImpl.java

```diff
@@ -325,7 +329,52 @@ private boolean isFailOnCCE() {
 		private void setLocalFailurePolicy(QueryFailurePolicy localFailurePolicy) {
 			this.localFailurePolicy = localFailurePolicy;
 		}
+
+		/**
+		 * check whether `input` can be used to call a function.
+		 * @param input the to be checked value
+		 * @return true if it can be used or if we do not know that yet
+		 */
+		protected boolean isAcceptableType(Object input) {
+			if (expectedClass != null && expectedClass.isAssignableFrom(input.getClass()) == false) {
+				onClassCastException(this, classCastException, input);
+				return false;
+			}
+			return true;
+		}
+
+		protected void detectExpectedClassFromCCE(ClassCastException e, Object input) {
+			if (expectedClass == null && canExtractTypeFromCCE) {
+				//detect expected class from CCE message
+				String message = e.getMessage();
+				if (message != null) {
+					Matcher m = cceMessagePattern.matcher(message);
+					if (m.matches()) {
+						String objectClassName = m.group(1);
+						String expectedClassName = m.group(2);
+						if (objectClassName.equals(input.getClass().getName())) {
+							try {
+								expectedClass = getClass().getClassLoader().loadClass(expectedClassName);
+								classCastException = e;
+							} catch (ClassNotFoundException e1) {
+								//the class cast exception message is invalid
+								if (Launcher.LOGGER.isDebugEnabled()) {
+									Launcher.LOGGER.debug("Unexpected ClassCastException message: \"" + message + "\"");
+								}
+								canExtractTypeFromCCE = false;
+							}
+						} else {
+							canExtractTypeFromCCE = false;
+						}
+					} else {
+						canExtractTypeFromCCE = false;
+					}
+				}
+			}
+		}
 	}
+	private static final Pattern cceMessagePattern = Pattern.compile("(\\S+) cannot be cast to (\\S+)");
+	private static boolean canExtractTypeFromCCE = true;
```

Comment:
Is it normal that canExtractTypeFromCCE cannot be set to true?

Reply:
the canExtractTypeFromCCE is here as fallback for the case, when somebody is using different java implementation whose CCE error message looks different. Only in such case the value is set to false and the detection of acceptable type from message is OFF.

> Is it normal that canExtractTypeFromCCE cannot be set to true?

So I see actually no case when such setter would be needed.